### PR TITLE
feat(rome_js_formatter): Function parameter & return type grouping

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR refines the formatting of function parameters and return type annotations by adopting Prettier's behaviour to first break the return type annotation in some situations before breaking the parentheses (see `shouldGroupCallArguments`). 

Implementing this change required to extract the shared method formatting logic into a `FormatMethodMember` helper struct.

## Test Plan

Verified that e.g. object types now break before the parentheses. 

File Based Average Prettier Similarity: 78.03% -> 78.10%
Line Based Average Prettier Similarity: 73.10% -> 73.19%

### CODE REVIEW

**Comment by MichaReiser** [comment]
!bench_formatter

File: crates/rome_js_formatter/src/js/classes/method_class_member.rs [hunk 1]
```diff
@@ -1,44 +1,217 @@
...
+                space(),
+                FormatMethodMember::from(node.clone())
+            ]
+        ]
+    }
+}
+
+declare_node_union! {
+    /// Formats the type parameters, parameters, and return type annotation of a method
+    pub(crate) FormatMethodMember =
```
Comment (ematipico) [inline crates/rome_js_formatter/src/js/classes/method_class_member.rs:31]: nit: `JsAnyFormatMethodMember` ? It's mostly to use one naming convention.

File: crates/rome_js_formatter/src/js/declarations/function_declaration.rs [hunk 2]
```diff
@@ -120,22 +135,92 @@ impl Format<JsFormatContext> for FormatFunction {
...
-        write!(f, [self.body().format()])
+        if let Some(body) = self.body()? {
+            write!(f, [space(), body.format()])?;
+        }
+
+        Ok(())
     }
 }
+
+// Grouping the parameters has the effect that the return type will break first.
```
Comment (ematipico) [inline crates/rome_js_formatter/src/js/declarations/function_declaration.rs:176]:
```suggestion
/// Grouping the parameters has the effect that the return type will break first.
```

File: crates/rome_js_formatter/src/js/declarations/function_declaration.rs [hunk 3]
```diff
@@ -120,22 +135,92 @@ impl Format<JsFormatContext> for FormatFunction {
...
+pub(crate) fn should_group_function_parameters(
+    type_parameters: Option<&TsTypeParameters>,
+    parameter_count: usize,
+    return_type: Option<SyntaxResult<TsAnyReturnType>>,
+    formatted_return_type: &mut Memoized<impl Format<JsFormatContext>, JsFormatContext>,
```
Comment (ematipico) [inline crates/rome_js_formatter/src/js/declarations/function_declaration.rs:181]: First time I see this pattern, and it seems weird (anti-pattern?). Do you foresee future cases where we might need to use this pattern?

File: crates/rome_js_formatter/src/js/declarations/function_declaration.rs [hunk 4]
```diff
@@ -120,22 +135,92 @@ impl Format<JsFormatContext> for FormatFunction {
...
+            _ => return Ok(false),
+        }
+    }
+
+    let result = if parameter_count != 1 {
+        false
+    } else {
+        // THIS is a hack that is necessary to avoid that the formatter doesn't insert a space
+        // between `)` and the `:` of the return type annotation IF there's a an inline comment
+        // after the last comment that has been written. This can be deleted once the comments refactor lands.
```
Comment (ematipico) [inline crates/rome_js_formatter/src/js/declarations/function_declaration.rs:211]: Is it worth adding a `TODO` with a link to the issue of the refactor?

File: crates/rome_js_formatter/src/js/declarations/function_declaration.rs [hunk 5]
```diff
@@ -120,22 +135,92 @@ impl Format<JsFormatContext> for FormatFunction {
...
+        let is_last_content_inline_comment = f.state().is_last_content_inline_comment();
+        f.state_mut().set_last_content_inline_comment(false);
+
+        let group = matches!(
+            return_type,
+            TsAnyReturnType::TsType(TsType::TsObjectType(_) | TsType::TsMappedType(_))
+        ) || formatted_return_type.inspect(f)?.will_break();
```
Comment (ematipico) [inline crates/rome_js_formatter/src/js/declarations/function_declaration.rs:218]: Could you update this [line](https://github.com/rome/tools/blob/main/crates/rome_formatter/src/buffer.rs#L489), this [line](https://github.com/rome/tools/blob/main/crates/rome_formatter/src/buffer.rs#L542-L545) - need to fix the markdown title too, it's commented, and add the try operator, please?

**Review by ematipico — APPROVED** [review]

File: crates/rome_js_formatter/src/js/declarations/function_declaration.rs (hunk 3 above)
Comment (MichaReiser) [reply crates/rome_js_formatter/src/js/declarations/function_declaration.rs:181]: Do you mean the need for passing in "Memoized" or the use of `impl` vs a generic type parameter? 

* Memoized: I need to inspect the content. There's no other way of doing this than passing in the memorized document
* `impl` vs generic type parameter. This are the same

**Pull request merged**